### ISSUE TITLE
Incorrectly handled clock_gettime64 syscall

### ISSUE THREAD

**Issue by igorsnunes** [docker/for-win#8326]
- [ ] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Information
Hi everyone,

While running my application on a i386 debian image (bullseye), I am constantly receiving an "Operation not permitted" for the clock_gettime system call. This error only happens when using a newer version of libc6 (2.31-1). And when this event happens, at some point, my application stops working.

Doing some investigation I figured out that newer versions of glibc, clock_gettime() syscalls falls back to clock_gettime64(). When using "strace" (to scan system calls) on my application, I can see that when clock_gettime64() is called, an EPERM is returned. This specific error code breaks the application.   The ploblem with this is:  glibc expects a ENOSYS, indicating that this syscall is not implemented by the kernel. If that happens, libc uses another implementation of clock_gettime, returning the correct value; if EPERM is returned instead, libc handles this return value as an error. 

I can bypass this issue by running the container with the “—privileged” flag, or creating a seccomp profile that has the following configuration:

"defaultAction": "SCMP_ACT_TRACE"

Which means: return ENOSYS as a default behavior, instead of EPERM.

The –privileged flag bypasses seccomp, and allow every syscall  to be handled by the kernel (and apparently, the kernel returns the correct code).

Question: why “clock_gettime64” is not being matched on any seccomp profile (including the default one, used by the engine)? The only way I managed to make this syscall returns ENOSYS, using seccomp profile, was enabling the defaultAction as SCMP_ACT_TRACE. And as far as I can see, this is not a good  practice; the correct action would be SCMP_ACT_ERRNO for default cases. See below the two approaches that I tried on my seccomp profile, and didn`t work:

Explicitly allowing clock_gettime64:
{
"names": ["clock_gettime64"],
              "action": "SCMP_ACT_ALLOW",
"args": [],
"comment": "",
"includes": {},
"excludes": {}
}

Explicitly setting the behavior of clock_gettime64 to TRACE:

{
"names": ["clock_gettime64"],
              "action": "SCMP_ACT_TRACE",
"args": [],
"comment": "",
"includes": {},
"excludes": {}
}

As shown here https://bugs.launchpad.net/ubuntu/+source/libseccomp/+bug/1868720 , this might be a problem related to older versions of libseccomp installed on the host. Is there a way to get this information from the host linux system used by Docker Desktop?

Please, let me know if I am missing something on my analysis.

Ps: Some documentation used for this analysis:
https://gitlab.alpinelinux.org/alpine/aports/-/issues/11774
https://lwn.net/Articles/795128/
https://docs.docker.com/engine/security/seccomp/
https://bugs.launchpad.net/ubuntu/+source/libseccomp/+bug/1868720

  - Windows Version: Windows 10 Pro, version 1903
  - Docker Desktop Version: 2.3.0.4 (46911)
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

### Steps to reproduce the behavior

  1. 

Compile the following code for 32 bits, i.e. "gcc -m32":

```
#include <stdio.h>
#include <time.h>
#include <fcntl.h>

int main () {
	struct timespec tp;
	
	if (clock_gettime(CLOCK_REALTIME, &tp) == -1) {
		perror("clock_gettime");
	}
	else {
		printf("clock_gettime success: %ld\n", tp.tv_nsec);
	}
	return 0;
}

```

  2. 
Run the binary on a docker image with a newer version of libc on a i386 environment. You can do:

```
docker run --entrypoint  bash   -v "C:\path_to_bin:/path_to_bin"  -it i386/debian:bullseye
```

**Comment by stephen-turner** [comment]
Thanks for the report, @igorsnunes. I found a seemingly relevant patch in version 19.03.9 of the upstream engine: https://github.com/moby/moby/compare/v19.03.8...v19.03.9. But we have engine 19.03.12 in Desktop 2.3.0.4 so maybe there's something else we need to do in our VM. We'll take a look.

**Comment by djs55** [comment]
Hi @igorsnunes -- we believe we need to upgrade the version of `libseccomp` bundled inside Docker Desktop. We currently link `libseccomp` statically into the `dockerd` binary. The simplest solution therefore is to bump the version inside the build environment -- which is in progress -- but unfortunately this is the build environment used to build other Linux packages and the change has knock-on effects for other architectures like `armhf` so it may take a while to fix.

We're also considering switching Docker Desktop to using a dynamically-linked `dockerd`, which would allow Desktop to bump the `libseccomp` version in our `Dockerfile` without worrying about effects on `armhf` (for now anyway!) However this is quite a big change to our build process too, so will take a while.

In summary
- we're working on it, but it might take a while
- I'll add a known issue to the docs referencing this issue
- we'll keep this issue open and keep you informed as we make progress.

Thanks again for your report. I was hoping there was a quick fix available but unfortunately I've failed to find one.

/lifecycle frozen

**Comment by raxvan** [comment]
Hello, just for information i found this issue while searching reasons why `clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &now_time);` returns incorrect time in `new_time.tv_nsec`. With the same code posted by @igorsnunes (except for `CLOCK_PROCESS_CPUTIME_ID`) i'm getting more or less random values in tv_nsec. I'm using docker engine with WSL2.
Running the image with `--privileged` solves the issue.

**Comment by igorsnunes** [comment]
Thanks @djs55  and @stephen-turner . I'll keep following your updates.

**Comment by microhobby** [comment]
Using a `Kernel v5.x` the issue does not occurs:

![image](https://user-images.githubusercontent.com/2633321/102727973-f9389880-4307-11eb-89ef-0eef988be95a.png)

These syscalls have been added if I'm not mistaken in `Kernel v5.1`. So, I hope that the next update of the WSL 2 Kernel, which is planned to use the `Kernel v5.4 LTS` should solve this.

**Comment by disconnect3d** [comment]
I am running into the same issue on MacOS with Docker for Desktop 3.2.2: the `clock_gettime64` syscall returns `EPERM`.

...and this can be workarounded with `--security-opt seccomp=unconfined` so its related to seccomp blocking the syscall. It seems that Docker [whitelisted this syscall in their default seccomp policy a year ago](https://github.com/moby/moby/commit/89fabf0f241292e929fbb2fbb794d58d8d697ab5), but for some reason this is not used in Docker for Desktop? Why?

**Comment by djs55** [comment]
I believe this issue is fixed on Docker Desktop 3.3.1 with the newer `runc`:

![Screenshot 2564-04-29 at 09 26 29](https://user-images.githubusercontent.com/198586/116527160-4c33e500-a8d2-11eb-8b3d-f0670d1ad254.png)
